conda_build/metadata.py

```diff
@@ -109,6 +109,31 @@ def remove_constructor(cls, tag):
 # used to avoid recomputing/rescanning recipe contents for used variables
 used_vars_cache = {}
 
+# Placeholders singletons for os/os.environ used to memoize select_lines().
+_selector_placeholder_os = object()
+_selector_placeholder_os_environ = object()
+
+
+@lru_cache(maxsize=200)
```

Comment:
These `maxsize=200` are more or less arbitrarily chosen.
The default is `128` IIRC, but I just wanted to specify some (reasonable) number such that the temptation to add `maxsize=None` later is lessened.
We may not want to use `maxsize=None` deliberately to avoid unbound resource usage downstream, e.,g. in case multiple recipes are rendered in the same process.

Reply:
Memoizing compiled patterns might give a slight advantage here

Reply:
As mentioned by @dholth below, the `re` module already caches compiled patterns.
In contrast to `re`'s functions, the wrappers here not only cache for the pattern but also the matched string.
This is because we have redundant `re.*(same_pattern, same_string)` calls due to the repeated re-processing mentioned in https://github.com/conda/conda-build/issues/5224#issue-2179165566 .
As such, the memoizations here not try to sidestep the recompilation of patterns but the actual regex search/matches.